metaflow/events.py

```diff
@@ -0,0 +1,83 @@
+from collections import namedtuple, OrderedDict
+from datetime import datetime
+
+
+MetaflowEvent = namedtuple("MetaflowEvent", ["name", "id", "timestamp", "type"])
+
+
+class MetaflowTrigger(object):
+    def __init__(self, _meta=[]):
+        _meta.sort(key=lambda x: x.get("timestamp") or float("-inf"), reverse=True)
+        self._meta = OrderedDict(
+            {
+                obj["name"]: MetaflowEvent(
+                    **{
+                        **obj,
+                        # Add timestamp as datetime. Guaranteed to exist for Metaflow
+                        # events - best effort for everything else.
+                        **(
+                            {"timestamp": datetime.fromtimestamp(obj["timestamp"])}
+                            if obj.get("timestamp")
+                            and isinstance(obj.get("timestamp"), int)
+                            else {}
+                        ),
+                    }
+                )
+                for obj in _meta
+            }
+        )
+
+    @property
+    def event(self):
+        """
+        Returns a `MetaflowEvent` object corresponding to the triggering event. If multiple events triggered the run, returns the latest event.
+        """
+        return next(iter(self.events), None)
+
+    @property
+    def events(self):
+        """
+        Returns a list of `MetaflowEvent` objects correspondings to all the triggering events.
+        """
+        return list(self._meta.values()) or None
+
+    @property
+    def run(self):
+        """
+        If the triggering event is a Metaflow run, returns the corresponding `Run` object. If multiple runs triggered the run, returns the latest run. `None` if the event is not a `Run`.
+        """
+        return next(iter(self.runs), None)
+
+    @property
+    def runs(self):
+        """
+        If the triggering events correspond to Metaflow runs, returns a list of `Run` objects. Otherwise returns `None`.
+        """
+        # to avoid circular import
+        from metaflow import Run
+
+        return [
+            Run(
+                obj.id[: obj.id.index("/", obj.id.index("/") + 1)],
+                _namespace_check=False,
+            )
+            for obj in filter(
+                lambda x: x.type in ["run"],
+                self._meta.values(),
+            )
+        ] or None
+
+    def __getitem__(self, key):
+        """
+        If triggering events are runs, `key` corresponds to the flow name of the triggering run. Returns a triggering `Run` object corresponding to the key. If triggering events are not runs, `key` corresponds to the event name and a `MetaflowEvent` object is returned.
+        """
+        if self.runs:
```

Comment:
for clarity, `return None` should not be in `else` but just there so it can serve as a clear fallback if the name doesn't exist. It's the current behavior but not clear if intended (it is in this case).

Reply:
will do